Fix First pressure production buildings not refreshing when the underlying resource they produce pressure for changes.

Skeletonxf opened the pull request:
A change in the number of native resources, the amount of pressure percent to produce (as before), or every 30 seconds, will trigger a recalculation. This ensures First buildings will produce native pressure of a resource added after the buildings, and the 30 second recalculation covers the edge case in which a planet is terraformed, changing the resource, but not the number of them. Now, even resource changing will update the pressure percentage produced within 30 seconds in the worst case.

It might be worth complementing this with a hook put onto every terraforming method that triggers a refresh of this calculation, but as this currently is, the 30 second checks will trigger a pressure update when everything else fails to.

Skeletonxf committed:
Add vanilla status effects

Skeletonxf committed:
ProduceNativePressurePct will now produce pressure for all native resources and update when the resources change

A change in the number of native resources, the amount of pressure percent to produce (as before), or every 30 seconds, will trigger a recalculation. This ensures First buildings will produce native pressure of a resource added after the buildings, and the 30 second recalculation covers the edge case in which a planet is terraformed, changing the resource, but not the number of them. Now, even resource changing will update the pressure percentage produced within 30 seconds in the worst case.

It might be worth complementing this with a hook put onto every terraforming method that triggers a refresh of this calculation, but as this currently is, the 30 second checks will trigger a pressure update when everything else fails to.

Skeletonxf commented:
I suppose I should mention that I also changed the hook to consider all native resources. This may or may not be interpreted as game changing or a bug fix depending on if you think the First buildings doing nothing with the pressure for non primary resources is a bug or intended behavior. I don't think vanilla *has* any ways of putting resources with pressure onto planets as native resources anyway. The Harvest card, Comets, Megafarms, Hydrogenators, Planetoid Tier 1 and 2 and Altars buildings all add native resources but none of them add pressure. Edit: I've been informed that the core system of the Expanse map generates planets with multiple native resources (with pressure).

Akronymus-git commented:
To add to the pressure of native resources: Fulrate adds some when imported.  Not sure how exactly they interact with the different native resources though.

DaloLorn commented:
I don't think Fulrate or Morphic Materials would be considered, no. Skeleton might know for sure, considering it's been in his mod for a while now.

Skeletonxf commented:
their hooks use the same `obj.modPressure(type, amount)` that the first buildings do, so I'd really have to test it in a game to be sure what's happening.

Skeletonxf commented:
It should be noted that producing pressure for native resources can result in weird scenarios where a planet with a primary resource that is tier 1 gains a primary resource that is tier 2 and produces pressure for both, despite First statuses setting the planet only to level 1. I fixed this oddity by [changing the primary resource identification code](https://github.com/Skeletonxf/star-ruler-2-mod-ce/commit/5ecbda5a2d1a4db64d6a83b740f6579d74f99d5e) but I am concerned this could be starting to stray from QoL into gameplay changing.

DaloLorn commented:
Considering that only mods allow you to add a resource (except food/water) without destroying the planet's existing resources... I do not see any meaningful gameplay effect. If `resources.as` already exists in the CP, I see precisely zero harm in doing this.

Skeletonxf commented:
I've now tested Fulrate (by extension Morphics as they use the same hook), and it still adds no additional pressure, so no changes there.

Skeletonxf committed:
Merge branch 'Main' of github.com:sol-oriens/SR2-Community-Patch into Main2

Skeletonxf committed:
Port CE primary resource sorting logic
>
> Complements pressure production buildings fix by together making sure that Ancient empires can always build the appropriate buildings on planets, and that pressure producing buildings work on all native resources the planet has

Skeletonxf commented:
This branch now contains the primary resource enhancements, and I've merged it to be up to date with Main again.

DaloLorn commented on SR2 Community Patch/scripts/server/components/Resources.as:
```diff
@@ -6,17 +6,73 @@ import statuses;
+		if (level > otherLevel)
+			return 1;
+
+		// sort as Resource::opCmp would if no tie break yet
```
I'm curious, why did you decide to copy the implementation of `Resource::opCmp(const Resource@)` when this was already how the game used to handle it?

Skeletonxf replied:
This is using the remaining part of the Resource compare method after the modified first part fails to tie break. If I called the vanilla compare method it would likely tie break on the fields I already compared with the refactored logic

DaloLorn replied:
Oh, I see.

DaloLorn commented:
Looks okay. Merging.
